pkg/server/server.go

```diff
@@ -0,0 +1,354 @@
+// Package server implements the HTTP and gRPC server used throughout Grafana
+// Agent.
+//
+// It is a grafana/agent-specific fork of github.com/weaveworks/common/server.
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	_ "net/http/pprof" // anonymous import to get the pprof handler registered
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/gorilla/mux"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/hashicorp/go-multierror"
+	"github.com/oklog/run"
+	otgrpc "github.com/opentracing-contrib/go-grpc"
+	"github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/weaveworks/common/middleware"
+	"golang.org/x/net/netutil"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/keepalive"
+)
+
+// Server wraps an HTTP and gRPC server with some common initialization.
+//
+// Unless instrumentation is disabled in the Servers config, Prometheus metrics
+// will be automatically generated for the server.
+type Server struct {
+	optsMut sync.Mutex
+	opts    Flags
+
+	// Listeners to use for connections. These will use TLS when TLS is enabled.
+	httpListener net.Listener
+	grpcListener net.Listener
+
+	updateHTTPTLS func(TLSConfig) error
+	updateGRPCTLS func(TLSConfig) error
+
+	HTTP       *mux.Router
+	HTTPServer *http.Server
+	GRPC       *grpc.Server
+}
+
+// New creates a new Server with the given config.
+//
+// r is used to register Server-specific metrics. If r is nil, no metrics will
+// be registered.
+//
+// g is used for collecting metrics from the instrumentation handlers, when
+// enabled. If g is nil, a /metrics endpoint will not be registered.
+func New(l log.Logger, r prometheus.Registerer, g prometheus.Gatherer, cfg Config) (srv *Server, err error) {
+	// TODO(rfratto): make a argument and remove from Config struct in v0.26.0.
```

Comment:
This func feels pretty long, thoughts on separating it out into more digestible sections?